### PR TITLE
Fixes to enable running otbn tests using vcs simulator

### PR DESCRIPTION
**Pull request by venkatk-ot**
This fixes couple of build issues seen when trying to run otbn tests using VCS ------------------------------------------------------------------------------ ------------------------------------------------------------------------------ Issue 1: XMR from class in SV package to module not permitted Error-[SV-LCM-HRP] Hierarchical reference in package ../src/lowrisc_dv_otbn_env_0.1/seq_lib/otbn_sec_wipe_err_vseq.sv, 18 otbn_env_pkg, "tb.dut.u_otbn_core.u_otbn_start_stop_control.StartSecureWipeImpliesRunning_A"
  Hierarchical reference from package 'otbn_env_pkg' into module 'tb' is not
  allowed.
  Remove hierarchical reference from the package.
------------------------------------------------------------------------------ Solution: Workaround as per synopsys suggestion in support article https://solvnetplus.synopsys.com/s/article/Cross-Module-or-Hierarchical-Reference-XMR-from-a-Class-Inside-SV-Package-1576091449555 Create a SV interface with no ports and instantiate it in TB. Add a function in this interface that can XMR signals/assertions in TB/DUT hierarchy Pass this interface handle as a virtual interface to the class in SV package using config_db ------------------------------------------------------------------------------ ------------------------------------------------------------------------------ Issue 2: Missing python library ninja causing otbn smoke test to fail at run time This library is needed to run the gen_binaries.py script. Here is the error
.......opentitan/scratch/master/otbn-sim-vcs/0.otbn_smoke/latest Traceback (most recent call last):
  File "...........opentitan/hw/ip/otbn/dv/uvm/gen-binaries.py", line 266, in <module>
    sys.exit(main())
  File "..........opentitan/hw/ip/otbn/dv/uvm/gen-binaries.py", line 177, in main
    return subprocess.run(cmd, cwd=args.destdir, check=False).returncode
  File "/usr/lib64/python3.8/subprocess.py", line 493, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/lib64/python3.8/subprocess.py", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib64/python3.8/subprocess.py", line 1720, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'ninja' make: *** [............opentitan/hw/dv/tools/dvsim/sim.mk:52: pre_run] Error 1 ------------------------------------------------------------------------------- Solution: Add ninja==1.11.1.1 requirement to pyproject.toml file and run the script ./util/sh/scripts/gen-python-requirements.sh to generate the updated python-requirements.txt file.
check in / commit python-requirements.txt and pyproject.toml file ------------------------------------------------------------------------------- -------------------------------------------------------------------------------

### CODE REVIEW

File: hw/ip/otbn/dv/uvm/env/otbn_path_if.sv [hunk 1]
```diff
@@ -0,0 +1,27 @@
+// Copyright lowRISC contributors (OpenTitan project).
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+// Bound into dut and used to help control assertions
+
+// VCS doesn't support accessing cross module components from a class inside SV package
+// See description and workaround suggestion below
+// https://solvnetplus.synopsys.com/s/article/Cross-Module-or-Hierarchical-Reference-XMR-from-a-Class-Inside-SV-Package-1576091449555
```
Comment (rswarbrick) [inline hw/ip/otbn/dv/uvm/env/otbn_path_if.sv:8]: To be fair to Synopsys, VCS is correct here! And the solvnetplus URL will only work for people who have an account on that site. The comment here can probably just be something like "This interface uses hierarchical references to control some assertions inside the design. This can then be used by virtual sequences (which can't use hierarchical references directly)"

The note about VCS is reasonable ("It worked on Xcelium, but we'd like it to work with VCS too!"), but probably belongs in the commit message.

File: hw/ip/otbn/dv/uvm/env/otbn_path_if.sv [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+// This file implements the workaround by providing a handle to a task in the module scope
+// that can be referenced by the class in the SV package scope using a virtual interface as an intermediary
+
+`define SSCTRL_HPATH tb.dut.u_otbn_core.u_otbn_start_stop_control
+`define SSCTRL_PATH  `"`SSCTRL_HPATH`"
+
+interface otbn_path_if();
```
Comment (rswarbrick) [inline hw/ip/otbn/dv/uvm/env/otbn_path_if.sv:16]: Can we rename this interface to make it slightly more specific (and "path" isn't quite right because it doesn't really describe `control_secwipe_running_assertion`).

Maybe `otbn_ssctrl_if`?

File: hw/ip/otbn/dv/uvm/env/otbn_path_if.sv [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  function string get_ssctrl_path();
+    return(`SSCTRL_PATH);
```
Comment (rswarbrick) [inline hw/ip/otbn/dv/uvm/env/otbn_path_if.sv:18]: Nit: Missing space, and I don't think we need the parens.

File: hw/ip/otbn/dv/uvm/env/otbn_path_if.sv [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  endfunction
+  function void control_secwipe_running_assertion(bit enable);
```
Comment (rswarbrick) [inline hw/ip/otbn/dv/uvm/env/otbn_path_if.sv:20]: Nit: Can you add a newline between the functions?

File: hw/ip/otbn/dv/uvm/env/seq_lib/otbn_sec_wipe_err_vseq.sv [hunk 5]
```diff
@@ -5,22 +5,19 @@
...
 class otbn_sec_wipe_err_vseq extends otbn_base_vseq;
   `uvm_object_utils(otbn_sec_wipe_err_vseq)
 
   `uvm_object_new
 
   function void control_secwipe_running_assertion(bit enable);
-    if (enable) $asserton(0, `SSCTRL_HPATH.StartSecureWipeImpliesRunning_A);
-    else $assertoff(0, `SSCTRL_HPATH.StartSecureWipeImpliesRunning_A);
+    cfg.path_vif.control_secwipe_running_assertion(enable);
   endfunction
```
Comment (rswarbrick) [inline hw/ip/otbn/dv/uvm/env/seq_lib/otbn_sec_wipe_err_vseq.sv:16]: I'd probably suggest dropping this function and using the interface directly at the four call sites.

File: hw/ip/otbn/dv/uvm/env/seq_lib/otbn_sec_wipe_err_vseq.sv [hunk 6]
```diff
@@ -5,22 +5,19 @@
...
 
   // Send a spurious secure wipe request and check we lock up
   task send_spurious_req();
-    string err_path = {`SSCTRL_PATH, ".secure_wipe_req_i"};
+    string err_path = {cfg.path_vif.get_ssctrl_path(), ".secure_wipe_req_i"};
```
Comment (rswarbrick) [inline hw/ip/otbn/dv/uvm/env/seq_lib/otbn_sec_wipe_err_vseq.sv:20]: Maybe it would make sense to fold this into the function, so we could end up with something like
```systemverilog
    string err_path = cfg.path_vif.resolve_path("secure_wipe_req_i");
```

File: hw/ip/otbn/dv/uvm/tb.sv [hunk 7]
```diff
@@ -258,6 +258,8 @@ module tb;
    .model_insn_cnt_i (model_insn_cnt)
   );
 
+  otbn_dut_if dut_if();
```
Comment (rswarbrick) [inline hw/ip/otbn/dv/uvm/tb.sv:261]: I don't think I understand this: did the name change to `otbn_path_if`?

File: pyproject.toml [hunk 8]
```diff
@@ -25,6 +25,7 @@ dependencies = [
   "libcst == 1.1.0", # 1.2.0+ needs Python 3.9+
   "lxml ~= 5.0",
   "mako==1.1.6",
+  "ninja==1.11.1.1",
```
Comment (rswarbrick) [inline pyproject.toml:28]: Since the Python stuff is a separate change can you pull it out into a separate commit? (The PR can hold both)

**Review by rswarbrick — COMMENTED** [review]
Thanks for the work: it's great if we can get the DV code working with more simulators!

Some detailed comments inline, but a structural point:
- Please split the two changes into two commits (possibly sharing the PR)
- Please give those commits more informative commit messages.

File: hw/ip/otbn/dv/uvm/tb.sv (hunk 7 above)
Comment (venkatk-ot) [reply hw/ip/otbn/dv/uvm/tb.sv:261]: My bad. I had originally named the new interface dut_if, then changed it to path_if since the entropy_src block had an interface named path_if. I missed to change it in this line. This is probably why the DV smoke test in the private lowRISC CI job failed. But I have no way to confirm other than fixing this and trying again since I am not able to access the logs
https://github.com/lowRISC/opentitan-private-ci/actions/runs/11172639469/job/31059422276

File: pyproject.toml (hunk 8 above)
Comment (venkatk-ot) [reply pyproject.toml:28]: done

**Review by venkatk-ot — COMMENTED** [review]
done

File: pyproject.toml (hunk 8 above)
Comment (venkatk-ot) [reply pyproject.toml:28]: done

File: hw/ip/otbn/dv/uvm/env/otbn_env.sv [hunk 9]
```diff
@@ -80,6 +80,10 @@ class otbn_env extends cip_base_env #(
                                                        cfg.rnd_vif)) begin
       `uvm_fatal(`gfn, "failed to get otbn_rnd_if handle from uvm_config_db")
     end
+    if (!uvm_config_db#(virtual otbn_path_if)::get(this, "", "path_vif",
+                                                       cfg.path_vif)) begin
```
[the remark this answers is not in the document]
Comment (venkatk-ot) [reply hw/ip/otbn/dv/uvm/env/otbn_env.sv]: done

File: hw/ip/otbn/dv/uvm/env/otbn_path_if.sv (hunk 3 above)
Comment (venkatk-ot) [reply hw/ip/otbn/dv/uvm/env/otbn_path_if.sv]: removed the parens in the interface construct . 
not sure if you meant the missing space in the control_secwipe_running_assertion function. i added one extra space to each of the two lines there to fix indent alignment to 2 spaces

File: hw/ip/otbn/dv/uvm/env/otbn_path_if.sv (hunk 4 above)
Comment (venkatk-ot) [reply hw/ip/otbn/dv/uvm/env/otbn_path_if.sv]: done

File: hw/ip/otbn/dv/uvm/env/seq_lib/otbn_sec_wipe_err_vseq.sv (hunk 5 above)
Comment (venkatk-ot) [reply hw/ip/otbn/dv/uvm/env/seq_lib/otbn_sec_wipe_err_vseq.sv]: done

File: hw/ip/otbn/dv/uvm/env/seq_lib/otbn_sec_wipe_err_vseq.sv (hunk 6 above)
Comment (venkatk-ot) [reply hw/ip/otbn/dv/uvm/env/seq_lib/otbn_sec_wipe_err_vseq.sv]: done

File: hw/ip/otbn/dv/uvm/tb.sv (hunk 7 above)
Comment (venkatk-ot) [reply hw/ip/otbn/dv/uvm/tb.sv]: fixed

File: pyproject.toml (hunk 8 above)
Comment (venkatk-ot) [reply pyproject.toml:28]: done

**Comment by venkatk-ot** [comment]
@rswarbrick , I have split the PR into two commits and addressed your comments. Please take a look

File: hw/ip/otbn/dv/uvm/env/otbn_ssctrl_if.sv [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+// This interface uses hierarchical references to control some assertions inside the design.
+// This can then be used by virtual sequences (which can't use hierarchical references directly)
+
+`define SSCTRL_HPATH tb.dut.u_otbn_core.u_otbn_start_stop_control
+`define SSCTRL_PATH  `"`SSCTRL_HPATH`"
+
+interface otbn_ssctrl_if;
+
+  function string resolve_path(string signal);
+    return({`SSCTRL_PATH, ".", signal});
```
Comment (rswarbrick) [inline hw/ip/otbn/dv/uvm/env/otbn_ssctrl_if.sv:15]: Nit: Probably better to use a more conventional style like this:
```systemverilog
    return {`SSCTRL_PATH, ".", signal};
```

File: hw/ip/otbn/dv/uvm/env/otbn_path_if.sv (hunk 3 above)
Comment (rswarbrick) [reply hw/ip/otbn/dv/uvm/env/otbn_path_if.sv]: Ah, sorry: I meant a space after "return". I'll leave a more helpful comment in a sec.

File: hw/ip/otbn/dv/uvm/env/otbn_ssctrl_if.sv (hunk 10 above)
Comment (venkatk-ot) [reply hw/ip/otbn/dv/uvm/env/otbn_ssctrl_if.sv]: done

**Comment by venkatk-ot** [comment]
@rswarbrick 
I have addressed your comment above and also updated the commit message based on the general guideline you shared in the other pull request.
Please take a look.

**Review by rswarbrick — APPROVED** [review]
Looks good to me! Thanks.
